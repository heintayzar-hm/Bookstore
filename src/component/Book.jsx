import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line react/prefer-stateless-function
class Book extends React.Component {
  constructor(props) {
    super();
    this.props = props;
  }

  render() {
    const {
      id, title, author, removeBook,
    } = this.props;
    return (
      <>
        <div>
          <article>
            <span>{id}</span>
            <h2>{title}</h2>
            <h5>{author}</h5>
            <button onClick={() => removeBook(id)} type="button">Remove</button>
          </article>
        </div>
      </>
    );
  }
}

Book.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  removeBook: PropTypes.func.isRequired,
};
export default Book;
