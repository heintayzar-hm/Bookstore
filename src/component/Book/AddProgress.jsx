import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line react/prefer-stateless-function
class AddProgress extends React.Component {
  constructor(props) {
    super();
    this.props = props;
  }

  render() {
    const {
      currentChapter, totalChapter, changeHandler, fromUpdateSection,
    } = this.props;
    return (
      <>
        <div className={(fromUpdateSection) ? 'grid p-3 grid-cols-1 gap-4' : 'grid p-3 grid-cols-2 gap-4'}>
          <div className="grid grid-cols-1 gap-5">
            <span className="font-important text-gray-50 text-xl font-semibold">Current Chapter</span>
            <input id="currentChapter" className="dark:bg-black dark:text-white p-2 px-3 border focus:outline-none rounded focus:border-blue border-solid tracking-widest text-brown" name="currentChapter" onChange={changeHandler} data-testid="book-title-input" type="number" min="0" max="10000" placeholder="Current Chapter" value={currentChapter} required />
          </div>
          <div className="grid grid-cols-1 gap-5">
            <span className="font-important text-gray-50 text-xl font-semibold">Total Chapter</span>
            <input id="totalChapter" className=" dark:bg-black dark:text-white p-2 px-3 border focus:outline-none rounded focus:border-blue border-solid tracking-widest text-brown" name="totalChapter" onChange={changeHandler} data-testid="author-title-input" type="number" min="0" max="10000" placeholder="Total Chapter" value={totalChapter} required />
          </div>
        </div>
      </>
    );
  }
}

AddProgress.propTypes = {
  currentChapter: PropTypes.number.isRequired,
  totalChapter: PropTypes.number.isRequired,
  changeHandler: PropTypes.func.isRequired,
  fromUpdateSection: PropTypes.bool.isRequired,
};
export default AddProgress;
