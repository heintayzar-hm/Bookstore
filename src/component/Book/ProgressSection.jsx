import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line react/prefer-stateless-function
class ProgressSection extends React.Component {
  constructor(props) {
    super();
    this.props = props;
  }

  render() {
    const {
      chapter, totalChapter,
    } = this.props;
    return (
      <>

        <div className="flex flex-wrap items-center justify-items-center gap-2">
          <div className=" w-[5.625rem] h-[5.625rem] p-[0.63em]"><div className="absolute rounded-[50%] w-[4.25em] h-[4.25em] border-[5px] border-solid border-blue rotate-45 border-l-[#e8e8e8]" /></div>
          <div className="flex flex-col">
            <div className="text-3xl font-important text-brown dark:text-white">
              {Math.floor((chapter / totalChapter) * 100)}
              %
            </div>
            <div className="text-sm opacity-50 font-important text-brown dark:text-white">Completed</div>
          </div>
        </div>
        <div className="grid grid-cols-1 font-normal gap-1">
          <span className="font-normal opacity-50 text-brown dark:text-white text-sm">CURRENT CHAPTER </span>
          <span className="tracking-tight font-normal">
            Chapter
            {' '}
            {chapter}
          </span>
          <button className="tracking-wider text-white font-normal bg-blue rounded" type="button">UPDATE PROGRESS</button>
        </div>
      </>
    );
  }
}

ProgressSection.propTypes = {
  chapter: PropTypes.string.isRequired,
  totalChapter: PropTypes.string.isRequired,
};
export default ProgressSection;
