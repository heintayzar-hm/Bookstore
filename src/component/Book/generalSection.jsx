import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line react/prefer-stateless-function
class GeneralSection extends React.Component {
  constructor(props) {
    super();
    this.props = props;
  }

  render() {
    const {
      id, title, author, removeBook, category, updateHandler,
    } = this.props;
    return (
      <>

        <div className="grid grid-cols-1">
          <span className="tracking-wide font-important font-bold opacity-50 text-sm text-brown">{category}</span>
          <h2 className="font-normal tracking-tight font-extrabold text-2xl">{title}</h2>
          <h5 className="text-lightblue text-sm font-normal tracking-wide">{author}</h5>
          <div className=" text-lightblue text-sm font-normal tracking-wide mt-5 grid grid-cols-2 gap-2 sm:flex sm:flex-wrap sm:items-center">
            <button type="button">Comments</button>
            <div className="w-[0.125em] h-5 bg-[#e8e8e8] mr-4 ml-4" />
            <button onClick={() => removeBook(id)} type="button">Remove</button>
            <div className="w-[0.125em] h-5 bg-[#e8e8e8] mr-4 ml-4" />
            <button type="button" onClick={() => updateHandler(id)}>Edit</button>
          </div>
        </div>

      </>
    );
  }
}

GeneralSection.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  category: PropTypes.arrayOf(PropTypes.string).isRequired,
  removeBook: PropTypes.func.isRequired,
  updateHandler: PropTypes.func.isRequired,
};
export default GeneralSection;
