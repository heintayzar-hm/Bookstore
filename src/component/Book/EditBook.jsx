/* eslint-disable no-alert */
import React from 'react';
import PropTypes from 'prop-types';
import { AiFillCloseCircle } from 'react-icons/ai';
import { connect } from 'react-redux';
import DropDown from '../AddBook/dropdown';
import AddProgress from './AddProgress';
// eslint-disable-next-line react/prefer-stateless-function
class EditBook extends React.Component {
  constructor(props) {
    super();
    this.props = props;
    this.state = { ...props };
  }

  submitHandler = (e) => {
    e.preventDefault();
    const { editHandler, updateHandler } = this.props;
    const { title, author } = this.state;
    // eslint-disable-next-line no-restricted-globals
    if (!confirm('Do you really want to update the book?')) {
      return;
    }
    if (title.trim() && author.trim()) {
      editHandler(this.state);
      updateHandler(-1);
      this.setState({
        author: '',
        title: '',
        category: [],
        currentChapter: 0,
        totalChapter: 0,
      });
    }
  }

  changeHandler = (e) => {
    e.preventDefault();
    this.setState((prevState) => {
      if (e.target.name === 'currentChapter' || e.target.name === 'totalChapter') {
        return ({
          ...prevState,
          [e.target.name]: +(e.target.value),
        });
      }
      return ({
        ...prevState,
        [e.target.name]: e.target.value,
      });
    });
  }

  dropdownChangeHandler = (value) => {
    this.setState((prevState) => ({
      ...prevState,
      category: [value.name],
    }));
  }

  render() {
    const {
      id, title, author, category, currentChapter, totalChapter,
    } = this.state;
    const { removeBook, updateHandler, categories } = this.props;
    return (
      <>
        <div className="fixed z-50 inset-0 transparent-bg overflow-scroll">
          <button type="button" onClick={() => updateHandler(-1)} className="absolute right-8 top-3">
            <AiFillCloseCircle className="w-8 h-8" />
          </button>
          <form onSubmit={this.submitHandler}>
            <h2 className="font-important tracking-wider text-[1.25rem] font-bold text-white text-center p-6 pb-3 text-3xl">
              Editing Book:
              {' '}
              {title}
            </h2>
            <div className="grid grid-cols-1 p-5 gap-11">
              <div className="grid grid-cols-2 gap-4">
                <div className="grid grid-cols-1 gap-5">
                  <span className="font-important text-gray-50 text-xl font-semibold">Title</span>
                  <input className="dark:bg-black dark:text-white p-2 px-3 border focus:outline-none rounded focus:border-blue border-solid tracking-widest text-brown" name="title" id="title" onChange={this.changeHandler} data-testid="book-title-input" type="text" placeholder="Book Title" value={title} required />
                </div>
                <div className="grid grid-cols-1 gap-5">
                  <span className="font-important text-gray-50 text-xl font-semibold">Author</span>
                  <input id="author" className=" dark:bg-black dark:text-white p-2 px-3 border focus:outline-none rounded focus:border-blue border-solid tracking-widest text-brown" name="author" onChange={this.changeHandler} data-testid="author-title-input" type="text" placeholder="Author" value={author} required />
                </div>
              </div>
              <div className="">
                <span className="font-important tracking-wider text-[1.25rem] font-bold text-white uppercase pb-3 ">Add Genre</span>
                <div className="grid grid-cols-10 pt-5 p-3">
                  <div className="col-span-9">
                    <DropDown
                      changeHandler={this.dropdownChangeHandler}
                      currentSelectName={categories.find((cat) => cat.name === category[0])}
                      categories={categories}
                    />
                  </div>
                  <button type="button" className="col-span-1" onClick={this.addHandler}>ADD</button>
                </div>
              </div>
              <div>
                <span className="font-important tracking-wider text-[1.25rem] font-bold text-white uppercase pb-3 ">Optional</span>
                <AddProgress
                  totalChapter={totalChapter}
                  currentChapter={currentChapter}
                  fromUpdateSection={false}
                  changeHandler={this.changeHandler}
                />
              </div>
              <div className="grid grid-cols-2 gap-6">
                <button className="p-3 bg-blue text-white rounded hover:bg-white hover:text-blue" type="submit">Edit Book</button>
                <button className="p-3 bg-blue text-white rounded hover:bg-white hover:text-blue" type="button" onClick={() => removeBook(id)}>Delete Book</button>
              </div>
            </div>
          </form>
        </div>
      </>
    );
  }
}
const mapStateToProps = (state) => ({
  // specify which state data to provide to the component
  categories: state.categories,
});
EditBook.propTypes = {
  removeBook: PropTypes.func.isRequired,
  editHandler: PropTypes.func.isRequired,
  updateHandler: PropTypes.func.isRequired,
  categories: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.any),
    PropTypes.string,
  ]).isRequired,
};
export default connect(mapStateToProps)(EditBook);
