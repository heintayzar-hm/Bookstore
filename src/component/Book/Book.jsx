import React from 'react';
import PropTypes from 'prop-types';
import GeneralSection from './generalSection';
import ProgressSection from './ProgressSection';
import EditBook from './EditBook';

// eslint-disable-next-line react/prefer-stateless-function
class Book extends React.Component {
  constructor(props) {
    super();
    this.props = props;
    this.state = {
    };
  }

  render() {
    const {
      id, title, author, removeBook, category, updateHandler,
      updateStateProps, editBook, totalChapter, currentChapter,
    } = this.props;
    return (
      <>
        <li className="border rounded p-8 border-[#e8e8e8]">
          <article className="grid grid-cols-2 gap-5 sm:grid-cols-3">
            <GeneralSection
              id={id}
              title={title}
              author={author}
              removeBook={removeBook}
              updateHandler={updateHandler}
              category={category}
            />
            <ProgressSection
              currentChapter={currentChapter}
              totalChapter={totalChapter}
              id={id}
            />
            {(updateStateProps)
              ? (
                <EditBook
                  id={id}
                  title={title}
                  author={author}
                  removeBook={removeBook}
                  editHandler={editBook}
                  totalChapter={totalChapter}
                  currentChapter={currentChapter}
                  updateHandler={updateHandler}
                  category={category}
                />
              )
              : null}
          </article>
        </li>
      </>
    );
  }
}

Book.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  removeBook: PropTypes.func.isRequired,
  updateHandler: PropTypes.func.isRequired,
  updateStateProps: PropTypes.bool.isRequired,
  author: PropTypes.string.isRequired,
  editBook: PropTypes.func.isRequired,
  totalChapter: PropTypes.number.isRequired,
  currentChapter: PropTypes.number.isRequired,
  category: PropTypes.arrayOf(PropTypes.string).isRequired,
};
export default Book;
