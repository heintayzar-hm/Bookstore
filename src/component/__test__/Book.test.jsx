import { describe, expect, test } from 'vitest';
import { render } from '@testing-library/react';
import Book from '../Book';

describe('Book', () => {
  test('renders the initial state correctly', () => {
    // Arrange

    // Act
    const { container } = render(
      <Book />,
    );
    // Assert
    expect(container).toMatchSnapshot();
  });

  test('Have Remove Button', () => {
    // Arrange

    // Act
    const { getAllByText } = render(
      <Book />,
    );
    // Assert
    expect(getAllByText('Remove').length).toBe(1);
  });
});
