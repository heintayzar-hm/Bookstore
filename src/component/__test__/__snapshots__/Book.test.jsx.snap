// Vitest Snapshot v1

exports[`Book > renders the initial state correctly 1`] = `
<div>
  <div>
    <article>
      <span />
      <h2 />
      <h5 />
      <button
        type="button"
      >
        Remove
      </button>
    </article>
  </div>
</div>
`;
